src/EditorFeatures/Core/Implementation/ChangeSignature/AbstractChangeSignatureCommandHandler.cs

```diff
@@ -53,37 +54,36 @@ public bool ExecuteCommand(ReorderParametersCommandArgs args, CommandExecutionCo
 
         private bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, CommandExecutionContext context)
         {
-            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-            if (document == null)
-            {
-                return false;
-            }
-
-            // TODO: reuse GetCommandState instead
-            var workspace = document.Project.Solution.Workspace;
-            if (!workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
+            using (context.OperationContext.AddScope(allowCancellation: true, FeaturesResources.Change_signature))
             {
-                return false;
-            }
+                var document = subjectBuffer.CurrentSnapshot.GetFullyLoadedOpenDocumentInCurrentContextWithChangesAsync(
+                    context.OperationContext).WaitAndGetResult(context.OperationContext.UserCancellationToken);
+                if (document == null)
+                {
+                    return false;
+                }
 
-            var supportsFeatureService = document.Project.Solution.Workspace.Services.GetService<IDocumentSupportsFeatureService>();
-            if (!supportsFeatureService.SupportsRefactorings(document))
-            {
-                return false;
-            }
+                // TODO: reuse GetCommandState instead
+                var workspace = document.Project.Solution.Workspace;
+                if (!workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
+                {
+                    return false;
+                }
```

Comment:
this is bad.  cheap code should be lifted to the front.

Reply:
same answer.

Reply:
Same reason.  The current code will block a user for potentially a large amount of time, only to then dismiss and not actually do what they asked for....   That's a bad experience.  

If you do the fast check first, the user isn't pointlessly blocked.  We do this all over the place in the ide, and it's just a normal and widespread code pattern found everywhere...

Reply:
Let's step back a bit.  Here we are getting Workspace instance off of document. Why is that? Can we get the workspace directly? Once we do that we don't need document in order to check `CanApplyChange`.

Reply:
I think we should unwrap some of the layers of extension methods here. Then we can easily move the check at the beginning of the method like so:

```
if (Workspace.TryGetWorkspace(subjectBuffer.CurrentSnapshot.TextBuffer.AsTextContainer(), out var workspace) && 
    workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
{
  return false;
}
```

Similarly in other cases. Perhaps we can also define TryGetWorkspace extension on snapshot, to avoid repeating `.TextBuffer.AsTextContainer()`.

Reply:
@CyrusNajmabadi I think that would address (some of) your concerns, correct?

Reply:
I made a PR that does that sort of thing tmat. Note: it's still necessary to have document in some cases. For example, some capabilities are document based.  This has always been true in the past and is becoming even more true in the future with the razor work.

Reply:
>@CyrusNajmabadi I think that would address (some of) your concerns, correct? 

My primary concerns are:

1. There aren't multiple documents floating around that people have to get right
2. Users do not have to unnecessarily wait long periods of time just to have the feature bail out

*Almost*any solution for that would likely be fine with me. I proposed a PR showing how I thought it could be done. There were legit concerns on it though. So I'm open to other approaches that satisfy the above. Thanks!

Reply:
> Note: it's still necessary to have document in some cases. For example, some capabilities are document based. 

Let's look at those separately then. I think going directly to the workspace where we can is a good change in any case.

Reply:
Note: 1 is not speculative. The original API design worked on documents we m which features would have, and which would then be passed along.  So it's a very real concern, esp as I'm not sure if we've actually written tests for anything here.

Reply:
@heejaechang Can you identify where we depend on document capabilities, rather then workspace capabilities?

Reply:
>      Let's look at those separately then. I think going directly to the workspace where we can is a good change in any case.

Sure. But those exist in this pr I believe. So I would like a solution there. 

Note: there should already be a buffer->workspace helper somewhere. 

Also note: I don't think this API should be snapshot based. I thin k it should be buffer based.  

The wait for loading might literally cause a change to the editor file. So you might get back a document for a previous snapshot. And you would then still not be able to apply things. 

Being buffer based says: wait for load, then give me the doc for the *latest* snapshot. 

I made that change in my PR.  You'll also note it cleans stuff up since all callers start with a buffer since that's the natural entry point currency.

Reply:
Just briefly looking around I found `VisualStudioDocumentSupportsFeatureService.SupportsRefactorings`, `SupportsCodeFixes`, etc.

> Aside: These take `Document` instance but do not actually use it as a snapshot. They only use `DocumentId`. Seems like the fact whether a document supports a service or not should not depend on the version of the document. 

Is the association of `DocumentId` to buffers and contained document relationships something that we need to wait for the full solution load to finish in order to know?

If so then we can't move these checks before the wait for full solution load.

Reply:
> Also note: I don't think this API should be snapshot based. I thin k it should be buffer based.

Yeah, makes sense. I have updated my comment above.

Reply:
In general, the association is extremely quick.  I'm not sure if it's guaranteed, but in practice I know I haven't been and to notice it, and I'm pretty perf sensitive.  This contrasts with full project/solution load, which is routinely over a minute at least for me

Reply:
> They only use DocumentId. Seems like the fact whether a document supports a service or not should not depend on the version of the document and it would be better to change these to take DocumentId

+1000 on doing that.  In general, it is highly desirable to have Workspace+DocId apis for your stuff for the exact reason.  They're cheap and lightweight. Documents are not bad, but it's nice to avoid them if you can!

Reply:
I am not sure why this is related to this PR but sure if we really want, we can do that in another PR. but I do not believe it is necessary and it is just nick picking. 

for example, things like VisualStudioDocumentSupportsFeatureService accepts Document and there is no reason that we can assume it can answer the question like SupportsRefactoring without full solution info except we know the implementation we have right now. if we believe, it will be the only implementation we need, then, I am not sure why we even have this abstraction?

same goes for most other ones. this seems just all micro-optimizations without any actual backing data that this will become an issue. and in a way, the issue was brought up. also, if it does happen, fixing is not hard as well. so I really don't get why this become such a high issue like "THIS MUST BE ROLL BACK!".

Reply:
if we really concern on perf like this in this much degree. I am really surprised that we did something like this (https://github.com/dotnet/roslyn/issues/33780) and checked in.

since perf implication of that is like x1000 times than what we are doing here since it is one-time cost for an explicit command invocation where the work we do the above happens for every single diagnostic reported. 

but sure, I mean if someone send PR to do the check to make the micro optimization, sure go ahead. I won't block it.

Reply:
These are not microoptimizations, and I'm not sure why you're characterizing then as such.  

Waiting for solution load can take more than a minute.  The PR changed from code where we would bail out quickly if it didn't apply to one where the user would have to wait a long time before bailing out.

Reply:
> so I really don't get why this become such a high issue

Because you ignored this issue when I brought it up in the PR and when I made sure my code change handled it. It's irresponsible.  

We don't **knowingly** regress perf scenarios and then say: it's ok, if there is data we'll fix it. 

We should be responsible and thoughtful and try to keep things high quality to the best of or ability.  That means not ignoring relevant feedback

Reply:
@heejaechang I brought up a question which I think is important for further work with partial loaded solution: Is the association of DocumentId to buffers and contained document relationships something that we need to wait for the full solution load to finish in order to know? Do we have a well defined state where we know all the DocumentIds but do not know Documents?